axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/PurchaseOrderServiceSupplychainImpl.java

```diff
@@ -365,4 +374,95 @@ public void updateToValidatedStatus(PurchaseOrder purchaseOrder) {
     purchaseOrder.setStatusSelect(PurchaseOrderRepository.STATUS_VALIDATED);
     purchaseOrderRepo.save(purchaseOrder);
   }
+
+  @Override
+  public String createShipmentCostLine(PurchaseOrder purchaseOrder) throws AxelorException {
+    PurchaseOrderService purchaseOrderService = Beans.get(PurchaseOrderService.class);
+    List<PurchaseOrderLine> purchaseOrderLines = purchaseOrder.getPurchaseOrderLineList();
+    ShipmentMode shipmentMode = purchaseOrder.getShipmentMode();
+    if (shipmentMode == null) {
+      return null;
+    }
+    Product shippingCostProduct = shipmentMode.getShippingCostsProduct();
+    if (shipmentMode.getHasCarriagePaidPossibility()) {
+      BigDecimal carriagePaidThreshold = shipmentMode.getCarriagePaidThreshold();
+      if (computeExTaxTotalWithoutShippingLines(purchaseOrder).compareTo(carriagePaidThreshold)
+          >= 0) {
+        String message = removeShipmentCostLine(purchaseOrder);
+        purchaseOrderService.computePurchaseOrder(purchaseOrder);
+        return message;
+      }
+    }
+    if (alreadyHasShippingCostLine(purchaseOrder, shippingCostProduct)) {
+      return null;
+    }
+    PurchaseOrderLine shippingCostLine = createShippingCostLine(purchaseOrder, shippingCostProduct);
+    purchaseOrderLines.add(shippingCostLine);
+    purchaseOrderService.computePurchaseOrder(purchaseOrder);
+    return null;
+  }
+
+  private PurchaseOrderLine createShippingCostLine(
```

Comment:
Change all private methods to protected, as the use of private is preventing any override from customer module